# Primeros pasos de Pandas

## Instalar e importar

Pandas es un paquete fácil de instalar. Abra su programa de terminal (para usuarios de Mac) o línea de comandos (para usuarios de PC) e instálelo usando cualquiera de los siguientes comandos:

conda install pandas **o** pip install pandas

Dado que estamos trabajando sobre un cuaderno de Jupyter, puede ejecutar esta celda:

In [4]:
#! pip install pandas
#! conda install pandas

El ! al principio ejecuta las celdas como si estuvieran en una terminal.

Para importar pandas, generalmente lo importamos con un nombre más corto ya que se usa mucho:

In [5]:
import pandas as pd

# Componentes principales de pandas: Series y DataFrames

Los dos componentes principales de los pandas son Series y DataFrame.

Una Series es esencialmente una columna y un DataFrame es una tabla multidimensional formada por una "colección de Series".

<img src="https://storage.googleapis.com/lds-media/images/series-and-dataframe.width-1200.png">


# Series

Las series son estructuras unidimensionales conteniendo un array de datos (de cualquier tipo "soportado por NumPy") y un array de etiquetas que van asociadas a los datos, llamado índice (index en la literatura en inglés):

Clase pandas.Series ( datos = Ninguno , índice = Ninguno , dtype = Ninguno , nombre = Ninguno , copia = Falso , fastpath = Falso )

Parámetros

** valor de tipo matriz de datos , iterable, dictado o escalar**

Contiene datos almacenados en Series. Si los datos son un dictado, se mantiene el orden de los argumentos.

** índice similar a una matriz o índice (1d) **

Los valores deben ser hash y tener la misma longitud que los datos . Se permiten valores de índice no únicos. El valor predeterminado será RangeIndex (0, 1, 2,…, n) si no se proporciona. Si los datos son similares a un dictado y el índice es Ninguno, las claves de los datos se utilizan como índice. Si el índice no es Ninguno, la Serie resultante se vuelve a indexar con los valores del índice.

** dtype str, numpy.dtype o ExtensionDtype, opcional**

Tipo de datos para la serie de salida. Si no se especifica, esto se deducirá de los datos . Consulte la guía del usuario para conocer más usos.

** nombre str, opcional**

El nombre para darle a la Serie.

** copy bool, predeterminado Falso**

Copie los datos de entrada.

In [6]:
import pandas as pd

d = {'a': 1, 'b': 2, 'c': 3}
ser = pd.Series(data=d, index=['a', 'b', 'c'])
ser

a    1
b    2
c    3
dtype: int64

In [7]:
ventas1 = pd.Series([15,12,21]) # c/indice implicito
print(ventas1)
ventas = pd.Series([15,12,21], index = ["Ene", "Feb", "Mar"]) # c/indice explicito
ventas

0    15
1    12
2    21
dtype: int64


Ene    15
Feb    12
Mar    21
dtype: int64

Los elementos de la serie pueden extraerse con el nombre de la serie y, entre corchetes, el índice (posición) del elemento:

In [8]:
ventas[0] #mostrar a traves del valor implicito (indice)

15

o con su etiqueta, si la tiene: 

In [9]:
ventas["Ene"] #mostrar a traves del valor explicito (etiqueta)

15

Las etiquetas que forman el índice no necesitan ser diferentes. Pueden ser de cualquier tipo (numérico, textos, tuplas...) siempre que sea posible aplicar la función hash sobre ellas.

hash : https://www.interactivechaos.com/python/function/hash

Es de destacar que el lazo entre una etiqueta y un valor se mantendrá salvo que lo modifiquemos explícitamente. Esto quiere decir que filtrar una serie o eliminar un elemento de la serie, por ejemplo, no va a modificar las etiquetas asignadas a cada valor.

Otro comentario importante es al respecto de la inmutabilidad del índice de etiquetas: aun cuando es posible asignar a una serie un nuevo conjunto de etiquetas a través del atributo index, intentar modificar un único valor del index va a devolver un error.

Al igual que ocurre con el array NumPy, una serie pandas solo puede contener datos de un mismo tipo. En la imagen anterior puede apreciarse el índice a la izquierda ("Ene", "Feb" y "Mar") y los datos a la derecha (15, 12 y 21). El tipo de la serie, accesible a través del atributo dtype (Se muestra en la parte inferior: int64), coincide con el tipo de los datos que contiene:

In [10]:
ventas.dtype

dtype('int64')

Podemos acceder a los objetos que contienen los índices y los valores a través de los atributos **index** y **values** de la serie, respectivamente:

In [11]:
ventas.index

Index(['Ene', 'Feb', 'Mar'], dtype='object')

In [12]:
ventas.values

array([15, 12, 21], dtype=int64)

Puede apreciarse en el ejemplo que el índice es de tipo "objeto".

La serie tiene, además, un atributo name, atributo que también encontramos en el índice. Una vez los hemos fijado, se muestran junto con la estructura al imprimir la serie:

In [13]:
ventas.name = "Ventas 2020"
ventas.name

'Ventas 2020'

In [14]:
ventas

Ene    15
Feb    12
Mar    21
Name: Ventas 2020, dtype: int64

In [15]:
ventas.index.name = "Meses"
ventas1.copy = ventas
ventas1


0    15
1    12
2    21
dtype: int64

In [16]:
print(ventas)

ventas1

Meses
Ene    15
Feb    12
Mar    21
Name: Ventas 2020, dtype: int64


0    15
1    12
2    21
dtype: int64

Obsérvese cómo, en este ultimo ejemplo, en la salida, tanto la serie como el índice se muestran con su nombre ("Ventas 2020" y "Meses", respectivamente).

El atributo **axes** nos da acceso a una lista con los ejes de la serie (solo contiene un elemento al tratarse de una estructura unidimensional):

In [17]:
ventas.axes

[Index(['Ene', 'Feb', 'Mar'], dtype='object', name='Meses')]

El atributo **shape** nos devuelve el tamaño de la serie:

In [18]:
ventas.shape

(3,)

Para ver mas atributos de las Series se puede consultar la documentación de pandas: https://pandas.pydata.org/pandas-docs/stable/reference/series.html

# DataFrames 

Crear DataFrames directamente en Python y es bastante útil cuando se prueban nuevos métodos y funciones que se encuentran en los documentos de pandas.

Hay muchas formas de crear un DataFrame desde cero, pero una gran opción es usar un simple **dict**.

Digamos que tenemos un puesto de frutas que vende manzanas y naranjas. Queremos tener una columna para cada fruta y una fila para cada compra del cliente. Para organizar esto como un diccionario para pandas, podríamos hacer algo como:

# class pandas.DataFrame(data=None, index=None, columns=None, dtype=None, copy=False)

Datos tabulares bidimensionales, de tamaño mutable y potencialmente heterogéneos.

La estructura de datos también contiene ejes etiquetados (filas y columnas). Las operaciones aritméticas se alinean en las etiquetas de fila y columna. Se puede considerar como un contenedor similar a un dict para los objetos de la serie. La estructura de datos primaria de los pandas.

**Parámetros**

**data: **matriz de datos (estructurada u homogénea), Iterable, dict o DataFrame

Dict puede contener series, matrices, constantes, clases de datos u objetos de tipo lista. Si los datos son un dictado, el orden de las columnas sigue el orden de inserción.

**index: **índice o como una matriz
Índice que se utilizará para el fotograma resultante. Se establecerá de forma predeterminada en RangeIndex si no hay información de indexación como parte de los datos de entrada y no se proporciona un índice.

**columns: **índice o tipo matriz
Etiquetas de columna que se utilizarán para el marco resultante. El valor predeterminado será RangeIndex (0, 1, 2,…, n) si no se proporcionan etiquetas de columna.

**dtype: **dtype, predeterminado Ninguno
Tipo de datos a forzar. Solo se permite un solo tipo d. Si es Ninguno, infiera.

**copy: **bool, por defecto Falso
Copie los datos de las entradas. Solo afecta la entrada DataFrame / 2d ndarray.

In [19]:
# Diccionario
datos = { 'manzanas' : [ 3 , 2 , 0 , 1 ], 'naranjas' : [ 0 , 3 , 7 , 2 ] }

Y luego páselo al constructor de Pandas DataFrame:

In [20]:
import pandas as pd

compras = pd.DataFrame( data = datos, index = [1, 2, 3, 4] )
compras

manzanas  naranjas
1         3         0
2         2         3
3         0         7
4         1         2

### ¿Cómo funcionó eso?

Cada elemento (clave, valor) en "datos" corresponde a una columna en el DataFrame resultante.

El índice de este DataFrame se nos dio en la creación como los números 0-3, pero también podríamos crear el nuestro cuando inicializamos el DataFrame.

Tengamos nombres de clientes como nuestro índice:

In [21]:
# Diccionario
datos = { 'manzanas' : [ 3 , 2 , 0 , 1 ], 'naranjas' : [ 0 , 3 , 7 , 2 ] }

compras = pd.DataFrame( datos , index = [ 'Luis' , 'Roberto' , 'Liliana' , 'David' ])   
compras

manzanas  naranjas
Luis            3         0
Roberto         2         3
Liliana         0         7
David           1         2

Las etiquetas de filas y de columnas -los índices- son accesibles a través de los atributos index y columns, respectivamente:

In [22]:
compras.index

Index(['Luis', 'Roberto', 'Liliana', 'David'], dtype='object')

In [23]:
compras.columns

Index(['manzanas', 'naranjas'], dtype='object')

La nomenclatura usada por pandas puede resultar un tanto confusa en lo que se refiere a los índices: tanto la estructura que contiene las etiquetas de filas como la que contiene las etiquetas de columnas son objetos de tipo Index ("índice", en español), pero, como se ha comentado, el índice de filas se denomina también index (aunque en minúsculas), y el de columna, columns.

Además, el nombre de "indice" se aplica normalmente a la referencia de un dato en una estructura según su posición. Por ejemplo, en la lista m = ["a", "b"], el índice del primer elemento es el número o valor que, añadido entre corchetes tras el nombre de la lista, nos permite acceder al elemento. Así, el índice del elemento "a" en la lista mencionada es 0, y el índice del elemento "b" es 1, lo que no es del todo coherente con el concepto de "índice" de una estructura pandas cuando lo especificamos explícitamente.

Para evitar esta confusión, hablaremos normalmente de "índices" (en plural) para referirnos a estas dos estructuras (de filas y columnas), de "índice" (en singular) para referirnos al índice de etiquetas del eje vertical, y de "índice de columnas" y de "índice de filas" siempre que sea necesario remarcar a cuál estamos refiriéndonos.

In [24]:
compras.index

Index(['Luis', 'Roberto', 'Liliana', 'David'], dtype='object')

In [25]:
compras.columns

Index(['manzanas', 'naranjas'], dtype='object')

El eje 0 es el correspondiente al índice de filas (eje vertical) y el eje 1 al índice de columnas (eje horizontal). Como puede verse en el ejemplo anterior ejemplo, ambos índices son de tipo "objeto" (ya se ha comentado que, concretamente, son objetos de tipo Index).

El atributo axes devuelve una lista con los ejes de la estructura (dos, al tratarse de una estructura bidimensional):

In [26]:
compras.axes

[Index(['Luis', 'Roberto', 'Liliana', 'David'], dtype='object'),
 Index(['manzanas', 'naranjas'], dtype='object')]

Al igual que ocurría con las series, los índices de filas y columnas son inmutables. Esto significa que, aunque podemos asignar un nuevo conjunto de datos (etiquetas) a ambas estructuras (index o columns), intentar modificar un único valor devolverá un error.

Tanto el índice de filas como el de columnas poseen el atributo name. Una vez fijado, se muestra al imprimir la estructura:

In [27]:
#	manzanas	naranjas
#Juno	3	0
#Robert	2	3
#Lily	0	7
#David	1	2

# Adiciona un nombre para el indice

compras.index.name = "Clientes"

In [28]:
#	manzanas	naranjas
#Juno	3	0
#Robert	2	3
#Lily	0	7
#David	1	2

# Adiciona un nombre para manzanas y naranjas

compras.columns.name = "Frutas"

In [29]:
compras

Frutas    manzanas  naranjas
Clientes                    
Luis             3         0
Roberto          2         3
Liliana          0         7
David            1         2

De forma semejante a como ocurría con las series, el atributo values de un dataframe nos permite acceder a los valores del dataframe, con formato array NumPy 2d:

In [30]:
compras.values

array([[3, 0],
       [2, 3],
       [0, 7],
       [1, 2]], dtype=int64)

Este array tendrá un tipo u otro en función de los tipos de las columnas del dataframe, acomodándose de forma que englobe a todos ellos.

Y un dataframe también tiene un atributo shape que nos informa de su dimensionalidad y del número de elementos en cada dimensión. En el siguiente ejemplo Podemos ver que el dataframe compras tiene 4 filas y 2 columnas:

In [31]:
compras.shape

(4, 2)

Para ver más información sobre DataFrame se puede consultar la documentación de pandas: https://pandas.pydata.org/pandas-docs/stable/reference/frame.html

# Creación de series

El constructor para la creación de una serie pandas es pandas.Series. Este constructor acepta tres parámetros principales:

- data: estructura de datos tipo array, iterable, diccionario o valor escalar que contendrá los valores a introducir en la serie.
- index: estructura tipo array con la misma longitud que los datos. Si este argumento no se añade al crear la serie, se agregará un índice por defecto formado por números enteros desde 0 hasta n-1, siendo n el número de datos.
- dtype: tipo de datos para la serie. Si no se especifica, se inferirá a partir de los datos.

Los valores del índice, como ya se ha comentado anteriormente, no tienen que ser necesariamente distintos aunque ciertas operaciones pueden generar un error si no soportan la posibilidad de tener índices duplicados.


### Utilizando una lista 

In [32]:
import pandas as pd
s = pd.Series([7,5,3])
s

0    7
1    5
2    3
dtype: int64

Al no haberse especificado un índice, se asigna uno automáticamente con los valores 0, 1 y 2.

Si repetimos esta instrucción especificando un índice:

In [33]:
s = pd.Series(data = [7,5,3], index = ["Ene", "Feb", "Mar"])
s

Ene    7
Feb    5
Mar    3
dtype: int64

Aquí vemos cómo el índice por defecto ha sido sustituido por el indicado. En este caso, la longitud del índice deberá coindicir con el número de elementos de la lista.

Los mismos comentarios podrían hacerse si, en lugar de una lista, hubiésemos partido de un array NumPy para crear la serie.

### Utilizando un diccionario

In [34]:
d = {"Ene":7, "Feb":5, "Mar":3 }
s = pd.Series(d)
s

Ene    7
Feb    5
Mar    3
dtype: int64

Aquí vemos cómo el constructor utiliza las claves como etiquetas del índice, y los valores del diccionario como valores de la serie.

Si incluimos el índice explícitamente en el constructor, los valores en la serie se tomarán en el orden en el que estén en el índice explícito. Además, si en éste hay valores que no pertenecen al conjunto de claves del diccionario, se añaden a la serie con un valor NaN:

In [35]:
d = {"Ene":7, "Feb":5, "Mar":3 }
s = pd.Series(d, index = ["Abr", "Mar", "Feb", "Ene"],dtype=int) #NaN
s

Abr    NaN
Mar    3.0
Feb    5.0
Ene    7.0
dtype: float64

En este ejemplo, hemos creado la serie especificando el índice que hemos formado dando la vuelta a las claves del diccionario ("Mar", "Feb" y "Ene") y hemos añadido a la lista de etiquetas el valor "Abr", que no pertenece al conjunto de claves del diccionario. Se ha añadido a la serie, pero se le ha asignado el valor NaN. Es precisamente la presencia de este valor lo que modifica el tipo de la serie a float.

### Utilizando un escalar

Si los datos se reducen a un escalar (no a una lista con un único elemento, sino a un sencillo escalar como 7 o 15.4) será necesario añadir el índice explícitamente. El número de elementos de la serie coincidirá con el número de elementos del índice, y el escalar será asignado como valor a todos ellos:

In [36]:
s = pd.Series(7, index = ["Ene", "Feb", "Mar"])
s

Ene    7
Feb    7
Mar    7
dtype: int64

# Creación de dataframes

El constructor de dataframes es "pandas.DataFrame". Acepta cuatro parámetros principales:

- data: estructura de datos ndarray (array NumPy), diccionario u otro dataframe
- index: índice a aplicar a las filas. Si no se especifica, se asignará uno por defecto formado por números enteros entre 0 y n-1, siendo n el número de filas del dataframe.
- columns: etiquetas a aplicar a las columnas. Al igual que ocurre con el índice de filas, si no se añade se asignará uno automático formado por números enteros entre 0 y n-1, siendo n el número de columnas.
- dtype: tipo a aplicar a los datos. Solo se permite uno. Si no se especifica, se infiere el tipo de cada columna a partir de los datos que contengan.

Los valores de los índices de filas y columnas no tienen por qué ser necesariamente distintos.

Veamos algunas de las estructuras a partir de las que es posible construir un dataframe:

### Utilizando un diccionario

In [37]:
import pandas as pd
elementos = { 
    "Numero atómico":[1, 6, 47, 88],
    "Masa atómica":[1.008, 12.011, 107.87, 226],
    "Familia":["No metal", "No metal", "Metal", "Metal"]
}
elementos

{'Numero atómico': [1, 6, 47, 88],
 'Masa atómica': [1.008, 12.011, 107.87, 226],
 'Familia': ['No metal', 'No metal', 'Metal', 'Metal']}

Y creamos el dataframe con él como primer argumento:

In [38]:
tabla_periodica = pd.DataFrame(elementos)
tabla_periodica

Numero atómico  Masa atómica   Familia
0               1         1.008  No metal
1               6        12.011  No metal
2              47       107.870     Metal
3              88       226.000     Metal

El dataframe se ha creado situando las claves del diccionario como etiquetas de columnas y las listas asociadas a cada clave como columnas del dataframe. Al no haber especificado un índice de filas, éste ha tomado valores por defecto (0, 1, 2 y 3).

A continuación repetimos la misma operación especificando las etiquetas tanto para filas como para columnas, utilizando los parámetros index y columns, respectivamente:

In [39]:
tabla_periodica = pd.DataFrame(elementos,
                               index = ["H", "C", "Ag", "Ra"],
                               columns = ["Familia", "Numero atómico", "Masa atómica"]
)
tabla_periodica

Familia  Numero atómico  Masa atómica
H   No metal               1         1.008
C   No metal               6        12.011
Ag     Metal              47       107.870
Ra     Metal              88       226.000

Recordemos que con el parámetro columns podemos especificar el orden en el que se mostrarán las columnas o incluso filtrar éstas (no incluyendo todas las etiquetas presentes en el diccionario como claves), pero no cambiar sus nombres. De hecho, ya se ha comentado que si alguna de las etiquetas incluidas en dicho argumento no apareciese en el conjunto de claves del diccionario, se crearía una columna con dicho nombre pero con todos sus valores fijados a NaN.

Si, en lugar de listas de datos como valores del diccionario, hubiesen sido arrays NumPy o series, el procedimiento habría sido exactamente el mismo.

### Utilizando un array Numpy

En el caso de partir de un array NumPy, si no se especifican las etiquetas de filas y columnas, se asignan las etiquetas por defecto:

In [40]:
import numpy as np

unidades_Datos = np.array([[2, 5, 3, 2],
                           [4, 6, 7, 2], 
                           [3, 2, 4, 1]])
unidades_Datos 

array([[2, 5, 3, 2],
       [4, 6, 7, 2],
       [3, 2, 4, 1]])

In [41]:
unidades = pd.DataFrame(unidades_Datos)
unidades

0  1  2  3
0  2  5  3  2
1  4  6  7  2
2  3  2  4  1

Las filas del array NumPy siguen siendo interpretadas como filas del dataframe.

Si especificamos las etiquetas de filas y columnas, el resultado es diferente:

In [42]:
unidades = pd.DataFrame(data = unidades_Datos, index = [2015, 2016, 2017], columns = ["Ag", "Au", "Cu", "Pt"])
unidades

Ag  Au  Cu  Pt
2015   2   5   3   2
2016   4   6   7   2
2017   3   2   4   1

### Utilizando diferentes diccionarios

También podemos partir de un conjunto de diccionarios, cada uno definiendo el contenido de lo que será una fila del dataframe:

In [43]:
unidades_2015 = {"Ag":2, "Au":5, "Cu":3, "Pt":2}
unidades_2016 = {"Ag":4, "Au":6, "Cu":7, "Pt":2}
unidades_2017 = {"Ag":3, "Au":2, "Cu":4, "Pt":1}

unidades = pd.DataFrame([unidades_2015, unidades_2016, unidades_2017],
                       index = [2015, 2016, 2017])
unidades

Ag  Au  Cu  Pt
2015   2   5   3   2
2016   4   6   7   2
2017   3   2   4   1

Los diccionarios deberán compartir el mismo conjunto de claves que se interpretarán como etiquetas de columnas. Si las etiquetas no coinciden, se crearán todas las columnas pero se asignarán NaN a los valores desconocidos:

In [44]:
unidades_2015 = {"Ag":2, "Au":5, "Cu":3, "Pt":2}
unidades_2016 = {"Ag":4, "Au":6, "Cu":7, "Pt":2}
unidades_2017 = {"Ag":3, "Pb":2, "Cu":4, "Pt":1}

unidades = pd.DataFrame([unidades_2015, unidades_2016, unidades_2017],
                       index = [2015, 2016, 2017])
unidades

Ag   Au  Cu  Pt   Pb
2015   2  5.0   3   2  NaN
2016   4  6.0   7   2  NaN
2017   3  NaN   4   1  2.0

En este ejemplo, el año 2017 tiene una clave, Pb, que no existe en los otros dos diccionarios. Y este mismo año carece de la clave Au que sí se encuentra en los otros dos. Vemos cómo los datos no coincidentes se han rellenado con NaN.

# Otro metodos de contrucción

- pandas.DataFrame.from_dict, crea un dataframe a partir de un diccionario de diccionarios o de secuencias tipo array:
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.from_dict.html

- pandas.DataFrame.from_records, que parte de una lista de tuplas o de arrays NumPy con un tipo estructurado:
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.from_records.html

# Inspección de la información 

Normalmente, una vez hemos cargado un bloque de datos en una serie o un dataframe, lo primero que haremos será inspeccionarlo para confirmar que los datos cargados son los esperados y que la lectura se ha realizado correctamente. Para esto tenemos los métodos **head**, **tail** y **sample**, con un comportamiento semejante en series y dataframes, que nos muestran un subconjunto de los datos cargados. Además, los métodos **describe** e **info** nos proporcionan información adicional sobre los datos. Veamos estos métodos por separado.

# El método head

Este método, pandas.Series.head para series y pandas.DataFrame.head para dataframes, devuelve los primeros elementos de la estructura (los primeros valores en el caso de una serie y las primeras filas en el caso de un dataframe). Por defecto, se trata de los 5 primeros elementos, pero podemos especificar el número que deseamos como argumento de la función. Por ejemplo, partamos de las siguientes estructuras:

In [45]:
import pandas as pd
entradas = pd.Series([11, 18, 12, 16, 9, 16, 22, 28, 31, 29, 30, 12],
                    index = ["ene", "feb", "mar", "abr", "may", "jun", "jul", "ago",
                             "sep", "oct", "nov", "dic"])
entradas

ene    11
feb    18
mar    12
abr    16
may     9
jun    16
jul    22
ago    28
sep    31
oct    29
nov    30
dic    12
dtype: int64

In [46]:
salidas = pd.Series([9, 26, 18, 15, 6, 22, 19, 25, 34, 22, 21, 14],
                    index = ["ene", "feb", "mar", "abr", "may", "jun", "jul", "ago",
                             "sep", "oct", "nov", "dic"])
salidas

ene     9
feb    26
mar    18
abr    15
may     6
jun    22
jul    19
ago    25
sep    34
oct    22
nov    21
dic    14
dtype: int64

In [47]:
almacen = pd.DataFrame({"entradas": entradas, "salidas": salidas})
almacen["neto"] = almacen.entradas - almacen.salidas # así se crea una nueva columna. 
print(almacen)
print(almacen.shape)

     entradas  salidas  neto
ene        11        9     2
feb        18       26    -8
mar        12       18    -6
abr        16       15     1
may         9        6     3
jun        16       22    -6
jul        22       19     3
ago        28       25     3
sep        31       34    -3
oct        29       22     7
nov        30       21     9
dic        12       14    -2
(12, 3)


En este ejemplo estamos mostrando todos los elementos de la estructura pues son apenas 12. En un caso real podemos estar hablando de miles o de millones.

Ahora, para mostrar apenas los primeros elementos de la estructura, ejecutamos el método head:

In [48]:
entradas.head()

ene    11
feb    18
mar    12
abr    16
may     9
dtype: int64

In [49]:
almacen.head()

entradas  salidas  neto
ene        11        9     2
feb        18       26    -8
mar        12       18    -6
abr        16       15     1
may         9        6     3

# El método tail 

Los métodos pandas.Series.tail (para series) y pandas.DataFrame.tail (para dataframes) son semejantes a los anteriores, pero muestran los últimos elementos de la estructura. Si no indicamos otra cosa como argumento, serán los 5 últimos elementos los que se muestren:

In [50]:
entradas.tail()

ago    28
sep    31
oct    29
nov    30
dic    12
dtype: int64

In [51]:
almacen.tail()

entradas  salidas  neto
ago        28       25     3
sep        31       34    -3
oct        29       22     7
nov        30       21     9
dic        12       14    -2

# El método sample

Sin embargo, es frecuente que los datos que hayamos leído estén ordenados según algún criterio, y que el bloque de datos mostrado por los métodos head o tail estén formados por datos muy parecidos. Y en ocasiones nos puede convenir ver datos aleatorios de nuestra estructura. Para esto podemos utilizar los métodos "pandas.Series.sample" para series y "pandas.DataFrame.sample" para dataframes. Al contrario que head o tail, el número de elementos devueltos por defecto es uno, por lo que, si deseamos extraer una muestra mayor, tendremos que indicarlo como primer argumento:

In [52]:
entradas.sample()

ago    28
dtype: int64

In [53]:
almacen.sample(5)

entradas  salidas  neto
may         9        6     3
jul        22       19     3
ago        28       25     3
ene        11        9     2
nov        30       21     9

# El método describe

El método describe devuelve información estadística de los datos del dataframe o de la serie (de hecho, este método devuelve un dataframe). Esta información incluye el número de muestras, el valor medio, la desviación estándar, el valor mínimo, máximo, la mediana y los valores correspondientes a los percentiles 25% y 75%.

No  Función         Descripción
----------------------------------------------------
1   recuento ()     Número de observaciones no nulas

2   suma ()         Suma de valores

3   media ()        Media de valores

4   mediana ()      Mediana de valores

5   modo ()         Modo de valores

6   std ()          Desviación estándar de los valores

7   min ()          Valor mínimo

8   max ()          Valor máximo

9   abs ()          Valor absoluto

10  prod ()         Producto de valores

11  cumsum ()       Suma acumulada

12  cumprod ()      Producto acumulativo



Siguiendo con el ejemplo visto en la sección anterior:

In [54]:
import pandas as pd
entradas = pd.Series([11, 18, 12, 16, 9, 16, 22, 28, 31, 29, 30, 12],
                    index = ["ene", "feb", "mar", "abr", "may", "jun", "jul", "ago",
                             "sep", "oct", "nov", "dic"])
salidas = pd.Series([9, 26, 18, 15, 6, 22, 19, 25, 34, 22, 21, 14],
                    index = ["ene", "feb", "mar", "abr", "may", "jun", "jul", "ago",
                             "sep", "oct", "nov", "dic"])
almacen = pd.DataFrame({"entradas": entradas, "salidas": salidas})
almacen["neto"] = almacen.entradas - almacen.salidas

print(almacen.describe())

       entradas    salidas       neto
count  12.00000  12.000000  12.000000
mean   19.50000  19.250000   0.250000
std     8.16311   7.641097   5.310795
min     9.00000   6.000000  -8.000000
25%    12.00000  14.750000  -3.750000
50%    17.00000  20.000000   1.500000
75%    28.25000  22.750000   3.000000
max    31.00000  34.000000   9.000000


El método acepta el parámetro percentiles conteniendo una lista (o semejante) de los percentiles a mostrar. También acepta los parámetros include y exclude para especificar los tipos de las características a incluir o excluir del resultado.

# El método info

El método info muestra un resumen de un dataframe, incluyendo información sobre el tipo de los índices de filas y columnas, los valores no nulos y la memoria usada: 

In [55]:
almacen.info() #NaN

<class 'pandas.core.frame.DataFrame'>
Index: 12 entries, ene to dic
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   entradas  12 non-null     int64
 1   salidas   12 non-null     int64
 2   neto      12 non-null     int64
dtypes: int64(3)
memory usage: 384.0+ bytes


Solo los dataframes tienen implementado este método.

# El método value_counts

Un método de las series pandas extremadamente útil es pandas.Series.value_counts. Este método devuelve una estructura conteniendo los valores presentes en la serie y el número de ocurrencias de cada uno. Estos valores se muestran en orden decreciente:

In [56]:
import numpy as np
s = pd.Series([3, 1, 2, 1, 1, 4, 1, 2, np.nan])
print(s)
s.value_counts()

0    3.0
1    1.0
2    2.0
3    1.0
4    1.0
5    4.0
6    1.0
7    2.0
8    NaN
dtype: float64


1.0    4
2.0    2
3.0    1
4.0    1
dtype: int64

Como puede apreciarse, por defecto no se incluyen los valores nulos. Este comportamiento puede modificarse haciendo uso del parámetro **dropna**:

In [57]:
s.value_counts(dropna = False)

1.0    4
2.0    2
3.0    1
4.0    1
NaN    1
dtype: int64

En lugar de devolver los valores distintos y el número de ocurrencias, este método también puede agrupar los datos en "bins" y devolver una lista de bins (indicando sus márgenes) con el número de valores en cada uno de ellos. Por ejemplo, si quisiéramos agrupar los valores de la serie anterior en dos bins podríamos hacerlo de la siguiente forma:

In [58]:
s.value_counts(bins = 10)

(0.996, 1.3]    4
(1.9, 2.2]      2
(2.8, 3.1]      1
(3.7, 4.0]      1
(1.3, 1.6]      0
(1.6, 1.9]      0
(2.2, 2.5]      0
(2.5, 2.8]      0
(3.1, 3.4]      0
(3.4, 3.7]      0
dtype: int64

Vemos que se han creados los dos bins, el primero conteniendo los valores entre 0.996 y 2.5 (intervalo abierto por la izquierda y cerrado por la derecha), bin en el que hay 6 valores, y el segundo conteniendo los valores entre 2.5 y 4 (intervalo también abierto por la izquierda y cerrado por la derecha), bin en el que hay 2 valores.

-------------------------------------  Selección de información ----------------------------------

- - Selección de datos en Series

Ya se ha comentado que una serie pandas consta de un array de datos y un array de etiquetas (el índice o index). Si al crear la serie no se ha especificado el índice, ya sabemos que se asignará uno implícito por defecto:

In [59]:
import pandas as pd

s = pd.Series([10, 20, 30, 40])
s

0    10
1    20
2    30
3    40
dtype: int64

Podemos seleccionar los valores haciendo referencia al índice asignado con la misma notación que en un diccionario (la llamada "notación corchetes" o "square bracket notation"):

In [60]:
print(s[0])
print(s[2])

10
30


Usando esta sintaxis, si no se ha especificado un índice explícito, los índices negativos no están permitidos.

Si se asignan índices de forma explícita:

In [61]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "d"])
s

a    10
b    20
c    30
d    40
dtype: int64

podemos seleccionar los elementos usando el índice explícito o el implícito:

In [62]:
print(s["a"],s[0])

10 10


In [63]:
print(s["d"],s[3])

40 40


Con esta sintaxis, sí está permitido hacer uso de índices negativos para referirnos a los elementos desde el final de la estructura.

Si los índices asignados son números enteros (al igual que las etiquetas del índice implícito), el índice implícito queda desactivado:

In [64]:
s = pd.Series([10, 20, 30, 40], index = [3, 2, 1, 0])
s

3    10
2    20
1    30
0    40
dtype: int64

In [65]:
s[3]

10

# Uso de rangos

Es posible seleccionar rangos de valores. De esta forma, si usamos un rango numérico en una serie en la que hemos definido un índice explícito:

In [66]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "d"])
s

a    10
b    20
c    30
d    40
dtype: int64

In [67]:
s[1:3]

b    20
c    30
dtype: int64

observamos que el rango se interpreta como haciendo referencia al índice implícito, y se incluyen los valores desde el primer índice incluido, hasta el último sin incluir.

Si no se incluye alguno de los límites, el comportamiento es el estándar en Python (si no se incluye el primer valor, se consideran todos los elementos desde el principio, y si no se incluye el último valor, se consideran todos los elementos hasta el final):

In [68]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "d"])
s[1:]

b    20
c    30
d    40
dtype: int64

In [69]:
s[:3]

a    10
b    20
c    30
dtype: int64

Si se utilizan los índices explícitos en el rango, el comportamiento es ligeramente diferente:

In [70]:
s["a":"c"]

a    10
b    20
c    30
dtype: int64

In [71]:
s[:"c"]

a    10
b    20
c    30
dtype: int64

In [72]:
s["b":]

b    20
c    30
d    40
dtype: int64

Una posible fuente de confusión viene derivada del hecho de que, usando rangos, es posible hacer referencia tanto a las etiquetas como a los índices numéricos: si utilizamos etiquetas, hacemos referencia a las etiquetas (por supuesto):

In [73]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "d"])

In [74]:
s["b":"d"]

b    20
c    30
d    40
dtype: int64

y, por tanto, si utilizamos números, hacemos referencia a los índices numéricos (¿por supuesto...?):

In [75]:
s[1:3]

b    20
c    30
dtype: int64

¿Y qué ocurre si nuestras etiquetas son números? Pues que siempre que usemos rangos con números estaremos haciendo referencia a los índices numéricos: no es posible hacer referencia a las etiquetas:

In [76]:
s = pd.Series([10, 20, 30, 40], index = [3, 2, 1, 0])
s

3    10
2    20
1    30
0    40
dtype: int64

In [77]:
s[1:3]

2    20
1    30
dtype: int64

Sin embargo, algo como

In [78]:
s[1]

30

devolverá el valor cuya etiqueta es 1 (si existe):

In [79]:
s = pd.Series([10, 20, 30, 40], index = [3, 2, 1, 0])
s

3    10
2    20
1    30
0    40
dtype: int64

o el valor cuya posición es 1 si dicha etiqueta no existe y el índice explícito no es numérico:

In [80]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "d"])
print(s)
s[1]

a    10
b    20
c    30
d    40
dtype: int64


20

Es debido a esto que existen los métodos loc e iloc que veremos poco más adelante. Dichos métodos hacen una referencia explícita a etiquetas o posiciones, respectivamente, eliminando cualquier duda al respecto de su interpretación.

Al igual que con los array NumPy, es posible indicar, no un elemento simple ni un rango, sino una lista de valores. Por ejemplo:

In [81]:
#Serie
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "d"])
print(s)
s[[3, 1, 2]]

a    10
b    20
c    30
d    40
dtype: int64


d    40
b    20
c    30
dtype: int64

En este ejemplo, la lista contiene los números 3 y 1, y son los valores correspondientes a estos índices -y en el orden especificado- los devueltos por la instrucción.

El resultado devuelto sigue siendo una serie pandas:

In [82]:
type(s[[3, 1]])

pandas.core.series.Series

Con esta notación, en el caso de que la serie tenga un índice explícito numérico, los valores de la lista se interpretan como haciendo referencia al índice explícito.

También se puede utilizar el metodo **get**

In [83]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "d"])
print(s)
s.get(1) 

a    10
b    20
c    30
d    40
dtype: int64


20

In [84]:
s.get("b")

20

# Método loc 

El método pandas.Series.loc permite seleccionar un grupo de elementos por etiquetas.

Como argumento de este método puede utilizarse una única etiqueta:

In [85]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "b"])
s

a    10
b    20
c    30
b    40
dtype: int64

In [86]:
s.loc["b"]

b    20
b    40
dtype: int64

En este caso el argumento se interpreta siempre como etiqueta del índice, nunca como posición en dicho índice aun cuando se pase un número entero que no pertenece al conjunto de etiquetas y pueda representar una posición válida:

In [87]:
s.loc[0] #Error

KeyError: 0

También podemos pasar al método una lista de etiquetas, en cuyo caso se extraen los valores correspondientes a dichas etiquetas y en el orden en el que se incluyen en la lista:

In [88]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "d"])
print(s)
s.loc[["d","a"]]

a    10
b    20
c    30
d    40
dtype: int64


d    40
a    10
dtype: int64

Otra opción es pasar al método un rango:

In [89]:
s["b":"d"]

b    20
c    30
d    40
dtype: int64

En este caso es importante recalcar que, tal y como se ve en la imagen anterior, el método va a devolver todos los elementos entre los límites indicados ambos incluidos.

# Método iloc

El método pandas.Series.iloc permite extraer datos de la serie a partir de los índices implícitos que éstos tienen asignados.

La opción más simple es utilizar como argumento un simple número entero (el primer elemento de la serie recibe el índice cero):

In [90]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "d"])
s

a    10
b    20
c    30
d    40
dtype: int64

In [92]:
#a    10
#b    20
#c    30
#d    40
#dtype: int64

s.iloc[1]

20

In [93]:
#a    10
#b    20
#c    30
#d    40
#dtype: int64

s.iloc[0]

10

In [95]:
#a    10
#b    20
#c    30
#d    40
#dtype: int64

s.iloc[3]

40

Si el número es negativo, hace referencia al final de la serie (en este caso, el último elemento recibe el índice -1) -y esto tanto si se ha especificado un índice explícito como si no-:

In [96]:
#a    10
#b    20
#c    30
#d    40
#dtype: int64

s.iloc[-1]

40

In [97]:
#a    10
#b    20
#c    30
#d    40
#dtype: int64

s.iloc[-4]

10

Una segunda opción es pasar como argumento una lista o array de números, en cuyo caso se devuelven los elementos que ocupan dichas posiciones en el orden indicado en la lista o array:

In [98]:
#a    10
#b    20
#c    30
#d    40
#dtype: int64

s.iloc[[2, 0]]

c    30
a    10
dtype: int64

También podemos incluir en esta lista números negativos, con la funcionalidad ya comentada:

In [99]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "d"])
#a    10
#b    20
#c    30
#d    40
#dtype: int64

s.iloc[[-2, 0]]

c    30
a    10
dtype: int64

Una tercera opción es usar como argumento un rango de números:

In [100]:
#a    10
#b    20
#c    30
#d    40
#dtype: int64

s.iloc[1:3]

b    20
c    30
dtype: int64

Como vemos en el ejemplo anterior, si el rango tiene la forma a:b, se incluyen todos los elementos desde aquel cuyo índice es a (incluido) hasta el que tiene el índice b (sin incluir).

Si no se especifica el primer valor, se consideran todos los elementos desde el principio de la serie:

In [101]:
#a    10
#b    20
#c    30
#d    40
#dtype: int64

s.iloc[:3]

a    10
b    20
c    30
dtype: int64

Y, si no se especifica el segundo valor, se consideran todos los elementos hasta el final de la serie:

In [102]:
#a    10
#b    20
#c    30
#d    40
#dtype: int64

s.iloc[2:]

c    30
d    40
dtype: int64

También pueden usarse valores negativos para indicar el comienzo y/o el final del rango:

In [103]:
#a    10
#b    20
#c    30
#d    40
#dtype: int64

s.iloc[1:-1]

b    20
c    30
dtype: int64

# Uso de arrays booleanos

Una muy interesante opción para seleccionar elementos de una serie pandas es usar arrays booleanos. Por ejemplo, partimos de la siguiente serie:

In [104]:
s = pd.Series([5, 2, -3, 7, 8, 4])
s

0    5
1    2
2   -3
3    7
4    8
5    4
dtype: int64

Podemos seleccionar un conjunto de valores de la misma haciendo referencia al nombre de la serie y, entre los corchetes, una lista o array de booleanos (también puede ser una serie de booleanos, como veremos un poco más adelante):

In [105]:
#0    5
#1    2
#2   -3
#3    7
#4    8
#5    4
#dtype: int64

s[[True, False, False, True, True, False]]

0    5
3    7
4    8
dtype: int64

En este caso hemos seleccionado los elementos cuyos índices son 0, 3 y 4, que son los índices que ocupan los booleanos True en la lista de booleanos usada (lista cuya longitud deberá ser igual a la longitud de la serie pues, de no ser así, se devuelve un error).

Esta lista o array de booleanos no tiene porqué ser especificada de forma explícita, puede ser el resultado de una expresión:

In [107]:
#0    5
#1    2
#2   -3
#3    7
#4    8
#5    4
#dtype: int64

print(type(s > 2))
s > 2

<class 'pandas.core.series.Series'>


0     True
1    False
2    False
3     True
4     True
5     True
dtype: bool

Aquí, hemos usado la expresión s > 2 para generar una serie pandas de booleanos, serie en la que los valores toman el valor True cuando el valor con el mismo índice de s toma un valor mayor estricto que 2.

Podemos entonces usar este resultado para extraer valores de la serie s (valores que serán aquellos mayores que 2):

In [108]:
import pandas as pd
s = pd.Series([5, 2, -3, 7, 8, 4])
s[s>2]

0    5
3    7
4    8
5    4
dtype: int64

Este mismo enfoque puede ser usado con los métodos pandas.Series.loc y pandas.Series.iloc ya vistos en las secciones anteriores con algún matiz adicional:

El método loc puede ser usado tanto con un array explícito de booleanos:

In [109]:
s.loc[[True, False, False, True, True, True]]

0    5
3    7
4    8
5    4
dtype: int64

como con una expresión que genera, por ejemplo, una serie pandas de booleanos:

In [110]:
s.loc[s>2]

0    5
3    7
4    8
5    4
dtype: int64

Sin embargo, el método iloc tiene un comportamiento ligeramente diferente. Puede ser usados con arrays explícitos de booleanos:

In [111]:
s.iloc[[True, False, False, True, True, True]]

0    5
3    7
4    8
5    4
dtype: int64

pero el uso de expresiones que generen una serie pandas de booleanos devuelve un error:

In [113]:
s.iloc[s>2]

NotImplementedError: iLocation based boolean indexing on an integer type is not available

Si el objeto que está generando la estructura de booleanos (s, en s > 2) fuese un array NumPy en lugar de tratarse de una serie pandas, sí sería posible usar el método .iloc. De esta forma, la expresión s > 2 genera, como hemos visto, una serie pandas, pero podemos extraer los valores con el atributo values, que genera un array numpy:

In [114]:
type((s>2).values)

numpy.ndarray

In [115]:
(s>2).values

array([ True, False, False,  True,  True,  True])

Si usamos esta expresión para realizar la selección en la serie original s, el resultado es ahora el correcto:

In [116]:
s.iloc[(s>2).values]

0    5
3    7
4    8
5    4
dtype: int64

Es por ello que pandas recomienda usar el método loc cuando trabajemos con selección basada en booleanos.

# Selección aleatoria

También podemos realizar una selección aleatoria a partir de una serie. El método pandas.Series.sample permite especificar o bien el número de elementos a extraer o bien la fracción del número total de elementos a extraer (parámetros **n** y **frac**, respectivamente), pudiendo especificar si la extracción se realiza con reemplazo o no (parámetro **replace**), los pesos a aplicar a cada elemento para realizar una extracción aleatoria ponderada (parámetro **weights**), y una semilla para el generador de números aleatorios que asegure la reproducibilidad de la extracción (parámetro **random_state**). Por ejemplo:

In [117]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "s"])

In [141]:
s.sample(2, random_state = 6)

a    10
s    40
dtype: int64

Hemos extraído 3 elementos, por defecto sin reemplazo, aplicando el valor 18 como semilla del generador de números aleatorios.

In [145]:
s.sample(frac = 0.6, random_state = 18)

s    40
b    20
dtype: int64

En este otro ejemplo hemos extraído el 60% de los valores de la serie original haciendo uso del parámetro frac.

Si no hay reemplazo, el número máximo de elementos que podemos extraer coincide con la longitud de la serie. Pero si la extracción la realizamos con reemplazo, podemos especificar cualquier número de elementos:

In [146]:
s.sample(10, random_state = 18, replace = True)

c    30
s    40
a    10
b    20
c    30
b    20
c    30
c    30
c    30
a    10
dtype: int64

# Método pop

El método pandas.Series.pop extrae y elimina un elemento de una serie cuyo índice se indica como argumento ...devolviendo un error en caso de que no exista::

In [147]:
s = pd.Series([1, 2, 3, 4])
s

0    1
1    2
2    3
3    4
dtype: int64

In [148]:
s.pop(1)

2

In [149]:
s

0    1
2    3
3    4
dtype: int64

In [151]:
s = pd.Series([1, 2, 3, 4])
try:
    s.pop(2)
    print(s)
except:
    print("Error")

0    1
1    2
3    4
dtype: int64


Si la serie tiene un índice explícito, el argumento de pop hará referencia a este índice:

In [153]:
s = pd.Series([10, 20, 30, 40], index = ["a", "b", "c", "d"])
s

a    10
b    20
c    30
d    40
dtype: int64

In [154]:
s.pop("a")

10

In [155]:
s

b    20
c    30
d    40
dtype: int64

y no al implícito, lo que devolvería un error:

In [156]:
try:
    s.pop()
except:
    print("Error")

Error


# - - Selección de datos en DataFrames

Desde un punto de vista semántico, un dataframe puede ser considerado semejante a un diccionario de series, en el que las claves son los nombres de las columnas y los valores, las columnas (que son series pandas). En este ejemplo:

In [157]:
ventas = pd.DataFrame({
    "Entradas":[41, 32, 56, 18],
    "Salidas":[17, 54, 6, 78],
    "Valoración": [66, 54, 49, 66],
    "Límite": ["No", "Si", "No", "No"],
    "Cambio": [1.43, 1.16, -0.67, 0.77]},
    index = ["Ene", "Feb", "Mar", "Abr"])
ventas

Entradas  Salidas  Valoración Límite  Cambio
Ene        41       17          66     No    1.43
Feb        32       54          54     Si    1.16
Mar        56        6          49     No   -0.67
Abr        18       78          66     No    0.77

Podemos utilizar la sintaxis de los diccionarios para seleccionar la columna Entradas. Puede verse en el ejemplo siguiente cómo dicha columna es extraída con tipo de serie pandas:

In [158]:
print(type(ventas["Entradas"]))

<class 'pandas.core.series.Series'>


In [159]:
ventas["Entradas"]

Ene    41
Feb    32
Mar    56
Abr    18
Name: Entradas, dtype: int64

Esto significa que podemos realizar una selección en dicho resultado para, por ejemplo, extraer el valor correspondiente a febrero:

In [160]:
ventas["Entradas"]["Feb"]

32

Sin embargo, la más que razonable opción de eliminar los corchetes que separan ambos índices y sustituirlos por una coma no funciona:

In [161]:
ventas["Entradas","Feb"] #Error

KeyError: ('Entradas', 'Feb')

Si, una vez seleccionada una columna, le asignamos una lista o array (o serie) de valores de la misma longitud, estamos modificando dicha columna del dataframe:

In [163]:
print(ventas.Entradas)
ventas["Entradas"] = [33, 25, 40, 12]
ventas

Ene    33
Feb    25
Mar    40
Abr    12
Name: Entradas, dtype: int64


Entradas  Salidas  Valoración Límite  Cambio
Ene        33       17          66     No    1.43
Feb        25       54          54     Si    1.16
Mar        40        6          49     No   -0.67
Abr        12       78          66     No    0.77

Si asignamos un único valor escalar, este se propaga por toda la columna:

In [164]:
ventas["Salidas"] = 1
ventas

Entradas  Salidas  Valoración Límite  Cambio
Ene        33        1          66     No    1.43
Feb        25        1          54     Si    1.16
Mar        40        1          49     No   -0.67
Abr        12        1          66     No    0.77

Si estuviésemos asignando un array cuya longitud no coincidiese con la de la columna (y no estuviésemos asignando un escalar), obtendríamos un error.

Si asignamos una serie pandas se consideran los índices del dataframe y de la serie, haciendo coincidir los valores cuyos índices sean los mismos en ambas estructuras (si dicha columna no existe, se crea). En el caso de que haya valores en la serie con índices que no se encuentren en el dataframe, se descartan. Y en el caso de que haya índices en el dataframe que no se encuentren en la serie, se asigna un valor NaN.

Así, en el siguiente ejemplo, estamos añadiendo una serie cuyos índices son "Ene", "Mar", "Abr" y "May". Es decir, la serie no tiene un valor para el índice "Feb" que sí se encuentra en el dataframe (se asigna un NaN), e incluye el índice "May" que no se encuentra en el dataframe y se descarta:

In [167]:
ventas = pd.DataFrame({
    "Entradas":[41, 32, 56, 18],
    "Salidas":[17, 54, 6, 78],
    "Valoración": [66, 54, 49, 66],
    "Límite": ["No", "Si", "No", "No"],
    "Cambio": [1.43, 1.16, -0.67, 0.77]},
    index = ["Ene", "Feb", "Mar", "Abr"])
ventas

ventas["Pérdidas"] = pd.Series([5, 4, 6, 8], index = ["Ene", "Mar", "Abr", "May"])
ventas

Entradas  Salidas  Valoración Límite  Cambio  Pérdidas
Ene        41       17          66     No    1.43       5.0
Feb        32       54          54     Si    1.16       NaN
Mar        56        6          49     No   -0.67       4.0
Abr        18       78          66     No    0.77       6.0

Los valores asignados pueden proceder del propio dataframe:

In [168]:
ventas["Ganancias"] = (ventas["Entradas"]*2) - (ventas["Valoración"]/10)
ventas

Entradas  Salidas  Valoración Límite  Cambio  Pérdidas  Ganancias
Ene        41       17          66     No    1.43       5.0       75.4
Feb        32       54          54     Si    1.16       NaN       58.6
Mar        56        6          49     No   -0.67       4.0      107.1
Abr        18       78          66     No    0.77       6.0       29.4

También podemos acceder a una columna con la llamada "notación punto":

In [169]:
ventas.Ganancias

Ene     75.4
Feb     58.6
Mar    107.1
Abr     29.4
Name: Ganancias, dtype: float64

Deberemos tener en cuenta que con esta notación no es posible crear nuevas columnas ni eliminarlas con la función del y que solo funcionará si el nombre de la columna no incluye espacios en blanco y no coincide con ninguna palabra reservada de Python.

### Uso de rangos

El uso de un rango numérico entre los corchetes realiza una selección de filas, lo que puede parecer una cierta incoherencia:

In [170]:
#Crear un arreglo con indice, columnas y valores de 0 - 17
import numpy as np
import pandas as pd
df = pd.DataFrame(np.arange(18).reshape([6, 3]),
                  index = ["a", "b", "c", "d", "d", "f"],
                  columns = ["A", "B", "C"])
df

A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
d  12  13  14
f  15  16  17

In [171]:
df[2:5]

A   B   C
c   6   7   8
d   9  10  11
d  12  13  14

El equipo de pandas lo justifica diciendo que esta sintaxis resulta extremadamente conveniente al tratarse de un tipo de selección frecuentemente usada. Esto es cierto, pero el hecho de que selecciones aparentemente semejantes (df[1,2], df[[1, 2]], df[1:3, 5], etc.) devuelvan un error no facilita su comprensión.

En todo caso, vemos en el ejemplo anterior que se devuelven las filas entre el primer valor del rango (incluido) y el último (sin incluir). También podríamos haber usado las etiquetas del índice:

In [172]:
df["b":"d"]

A   B   C
b   3   4   5
c   6   7   8
d   9  10  11
d  12  13  14

In [173]:
df[:2]

A  B  C
a  0  1  2
b  3  4  5

In [174]:
df[:"c"]

A  B  C
a  0  1  2
b  3  4  5
c  6  7  8

Si, al realizar la selección, situamos entre los corchetes una lista de etiquetas, estaremos seleccionando columnas en el orden en el que aparecen en la lista y con formato dataframe:

In [175]:
df = pd.DataFrame(np.arange(18).reshape([6, 3]),
                  index = ["a", "b", "c", "d", "e", "f"],
                  columns = ["A", "B", "C"])
df

A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17

In [176]:
print(type(df[["C", "A"]]))
df[["C","A"]]

<class 'pandas.core.frame.DataFrame'>


C   A
a   2   0
b   5   3
c   8   6
d  11   9
e  14  12
f  17  15

También es posible extraer de forma segura una columna de un dataframe usando el método pandas.DataFrame.get. Éste extrae la columna indicada devolviendo un valor alternativo (por defecto None) si dicha columna no existe:

In [177]:
df = pd.DataFrame(np.arange(18).reshape([6, 3]),
                  index = ["a", "b", "c", "d", "e", "f"],
                  columns = ["A", "B", "C"])
df

A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17

In [178]:
df.get("A")

a     0
b     3
c     6
d     9
e    12
f    15
Name: A, dtype: int32

In [181]:
df.get("D")

# El método LOC

Al igual que ocurre con las series, el método pandas.DataFrame.loc permite seleccionar un conjunto de filas y columnas por etiquetas. Este método acepta diferentes argumentos. Para probarlos, partamos del siguiente dataframe:

In [182]:
df = pd.DataFrame(np.arange(18).reshape([6, 3]),
                  index = ["a", "b", "c", "d", "e", "f"],
                  columns = ["A", "B", "C"])
df

A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17

### Uso con etiqueta simple

El primer escenario lo encontramos cuando usamos este método indicando una única etiqueta. En este caso estamos seleccionando la fila cuya etiqueta se indique:

In [183]:
df.loc["c"]

A    6
B    7
C    8
Name: c, dtype: int32

El resultado es una serie pandas con las etiquetas de columnas del dataframe original como índice.

Es necesario mencionar que el argumento será siempre interpretado como etiqueta, aun cuando pueda estar representando un índice válido: 

In [184]:
df = pd.DataFrame(np.arange(12).reshape([4, 3]),
                  index = [1, 3, 0, 4],
                  columns = ["A", "B", "C"])
df

A   B   C
1  0   1   2
3  3   4   5
0  6   7   8
4  9  10  11

In [185]:
df.loc[0]

A    6
B    7
C    8
Name: 0, dtype: int32

Por supuesto, si dicha etiqueta no existe, se devuelve un error (nuevamente, aun cuando la etiqueta sea un número que pueda estar representando un índice válido.)

Si pasamos a loc una lista de etiquetas, estaremos extrayendo las filas cuyas etiquetas se indican, y en el orden en el que aparezcan en la lista:

In [186]:
df = pd.DataFrame(np.arange(18).reshape([6, 3]),
                  index = ["a", "b", "c", "d", "e", "f"],
                  columns = ["A", "B", "C"])
df

A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17

In [187]:
df.loc[["c","a","e"]]

A   B   C
c   6   7   8
a   0   1   2
e  12  13  14

Al contrario de lo que ocurre cuando solo indicamos una etiqueta, el resultado es un dataframe. Y lo es aún cuando la lista contenga un único elemento:

In [188]:
print(type(df.loc[["c"]]))
df.loc[["c"]]

<class 'pandas.core.frame.DataFrame'>


A  B  C
c  6  7  8

Otra opción es utilizar rangos limitados por etiquetas. De esta forma, si continuamos con el mismo ejemplo:

In [189]:
df.loc["b":"d"]

A   B   C
b  3   4   5
c  6   7   8
d  9  10  11

Obsérvese que la selección incluye todas las filas incluyendo las dos de los extremos del rango.

# Extracción de filas y columnas

En los ejemplos vistos hasta ahora estamos extrayendo una o varias filas para todas las columnas. En posible, por supuesto, especificar qué filas y qué columnas exactas queremos extraer. Así, si utilizamos una única etiqueta para indicar la fila, y una única etiqueta para indicar la columna, separadas por una coma, estaremos extrayendo un único valor:

In [190]:
#	A	B	C
#a	0	1	2
#b	3	4	5
#c	6	7	8
#d	9	10	11
#e	12	13	14
#f	15	16	17

import numpy as np
import pandas as pd
df = pd.DataFrame(np.arange(18).reshape([6, 3]),
                  index = ["a", "b", "c", "d", "e", "f"],
                  columns = ["A", "B", "C"])
print(df)

df.loc["a", "C"]

    A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17


2

Podemos sustituir una de las dos etiquetas por el símbolo de dos puntos (:), lo que supondrá seleccionar todos los elementos de ese eje:

In [191]:
df.loc[: ,"A"] # después de la coma van las columnas, antes es las filas. Igual a los dos lados de los dos puntos. 

a     0
b     3
c     6
d     9
e    12
f    15
Name: A, dtype: int32

Esto supone que, por ejemplo, las dos expresiones siguientes devuelven el mismo resultado:

In [192]:
df.loc["b"]

A    3
B    4
C    5
Name: b, dtype: int32

In [193]:
df.loc["b", :]

A    3
B    4
C    5
Name: b, dtype: int32

Los métodos vistos pueden combinarse. Podemos, por ejemplo, seleccionar la intersección de las filas e y c (en este orden) y la columna B:

In [194]:
df = pd.DataFrame(np.arange(18).reshape([6, 3]),
                  index = ["a", "b", "c", "d", "e", "f"],
                  columns = ["A", "B", "C"])
print(df)

df.loc[["e", "c"], "B"]

    A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17


e    13
c     7
Name: B, dtype: int32

# El método iloc

El método pandas.DataFrame.iloc permite realizar selecciones por posición. Tal y como cabría esperar, pueden utilizarse diferentes tipos de argumentos que determinan qué elementos se están extrayendo.

En este primer caso cuando hacemos uso un número entero estamos seleccionando la fila cuyo índice se indica:

In [195]:
df = pd.DataFrame(np.random.randint(0, 10, 18).reshape([6, 3]), 
               index = ["a", "b", "c", "d", "e", "f"],
               columns = ["A", "B", "C"])
df

A  B  C
a  8  3  8
b  2  0  6
c  0  7  1
d  3  3  1
e  9  9  8
f  7  8  1

In [196]:
df.iloc[2]

A    0
B    7
C    1
Name: c, dtype: int32

El número indicado siempre será tratado como posición, y no como etiqueta:

In [197]:
import pandas as pd
import numpy as np
df = pd.DataFrame(np.arange(12).reshape([4, 3]),
                 index = [3, 2, 1, 0],
                 columns = ["A", "B", "C"])
df

A   B   C
3  0   1   2
2  3   4   5
1  6   7   8
0  9  10  11

In [198]:
df.iloc[3]

A     9
B    10
C    11
Name: 0, dtype: int32

Si el número es negativo, hace referencia al final del dataframe. Así, siguiendo con este último ejemplo:

In [199]:
df.iloc[-1]

A     9
B    10
C    11
Name: 0, dtype: int32

Si utilizamos como argumento una lista o array de números, estamos extrayendo las filas cuyos índices son los elementos del mismo, y en el orden en el que aparecen en él:

In [200]:
df = pd.DataFrame(np.random.randint(0, 10, 18).reshape([6, 3]), 
               index = ["a", "b", "c", "d", "e", "f"],
               columns = ["A", "B", "C"])
df

A  B  C
a  5  8  5
b  8  8  4
c  2  1  1
d  8  5  5
e  4  6  0
f  8  1  1

In [201]:
df.iloc[[3, 1]]

A  B  C
d  8  5  5
b  8  8  4

En el ejemplo anterior, estamos extrayendo las filas cuyos índices son 3 y 1, y extrayéndolas en este mismo orden.

Si alguno de los índices es negativo, hará referencia al final de la lista.

Si utilizamos un rango de números, como en el siguiente ejemplo en el que indicamos como argumento 2:4, estamos extrayendo las filas cuyos índices van de la primera cifra del rango incluida (2 en el ejemplo) hasta la última cifra sin incluir (4 en el ejemplo):

In [202]:
df = pd.DataFrame(np.random.randint(0, 10, 18).reshape([6, 3]), 
               index = ["a", "b", "c", "d", "e", "f"],
               columns = ["A", "B", "C"])
print(df)

df.iloc[2:4]

   A  B  C
a  8  8  2
b  6  2  1
c  2  1  7
d  5  3  4
e  0  7  1
f  0  0  7


A  B  C
c  2  1  7
d  5  3  4

Como suele ser habitual, si no se especifica el primer valor, se consideran las filas desde la primera. Y si no se especifica el último valor, se consideran las filas hasta la última (incluida):

In [203]:
df.iloc[:3]

A  B  C
a  8  8  2
b  6  2  1
c  2  1  7

In [204]:
df.iloc[4:]

A  B  C
e  0  7  1
f  0  0  7

También pueden usarse valores negativos para especificar el comienzo o el final del rango.

# Extracción de filas y columnas

Si añadimos un segundo argumento, estaremos haciendo referencia al índice de columna:

In [205]:
df = pd.DataFrame(np.random.randint(0, 10, 18).reshape([6, 3]), 
               index = ["a", "b", "c", "d", "e", "f"],
               columns = ["A", "B", "C"])
print(df)

df.iloc[3, 1] #[F,C]

   A  B  C
a  8  0  9
b  1  6  1
c  6  6  3
d  7  4  4
e  0  7  4
f  8  2  6


4

En el ejemplo anterior estamos extrayendo el valor correspondiente a la fila cuyo índice es 3 y a la columna cuyo índice es 1.

#  Selección con índices y etiquetas simultáneamente

En ocasiones nos encontraremos con que resultaría de utilidad poder realizar selecciones mezclando etiquetas e índices, y los métodos vistos, loc e iloc, solo permiten el uso de etiquetas o de índices, respectivamente. Para poder mezclar ambos tipos de referencias podemos recurrir a los métodos pandas.Index.get_loc y pandas.Index.get_indexer, métodos asociados a los índices de un dataframe:

El primero, get_loc, devuelve el índice de la etiqueta que se adjunte como parámetro. El segundo, get_indexer, devuelve un array con los índices de las etiquetas que se adjunten en forma de lista como parámetro. Por ejemplo, partimos del siguiente dataframe:

In [206]:
df = pd.DataFrame(np.arange(18).reshape([6, 3]),
                 index = ["a", "b", "c", "d", "e", "f"],
                 columns = ["A", "B", "C"])
df

A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17

Si aplicamos los métodos comentados al índice de columnas haciendo referencia a etiquetas de columnas, obtenemos los siguientes resultados:

In [207]:
df.columns.get_loc("B")

1

In [208]:
df.columns.get_indexer(["A", "C"])

array([0, 2], dtype=int64)

En el primer caso hemos pasado la etiqueta "B" y el método ha devuelto su índice (1). En el segundo caso hemos pasado una lista de etiquetas y hemos obtenido un array con sus índices.

Si ejecutamos estos métodos en el índice de filas:

In [209]:
df.index.get_loc("d")

3

In [210]:
df.index.get_indexer(["c", "e"])

array([2, 4], dtype=int64)

obtenemos resultados semejantes.

Ahora que sabemos cómo convertir etiquetas en sus índices equivalentes, podemos seleccionar datos de un dataframe mezclando etiquetas e índices si convertimos las etiquetas y utilizamos el método iloc ya visto. Por ejemplo, si quisiéramos extraer del anterior dataframe el dato que ocupa la fila "c" y la columna de índice 2, podríamos conseguirlo del siguiente modo:

In [211]:
df = pd.DataFrame(np.arange(18).reshape([6, 3]),
                 index = ["a", "b", "c", "d", "e", "f"],
                 columns = ["A", "B", "C"])
print(df)
df.iloc[df.index.get_loc("c"), 2] #Obtener el dato la [fila "c", columna 2] equivalente a df.iloc[2,2]

    A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17


8

O si deseásemos obtener de las filas 5 y 3 (en este orden) los valores correspondientes a las columnas C y A (en este orden), podríamos hacerlo con la siguiente expresión:

In [212]:
print(df)
df.iloc[[5, 3], df.columns.get_indexer(["C", "A"])]

    A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17


C   A
f  17  15
d  11   9

# Uso de listas de booleanos

Otro método especialmente útil para la selección es el uso de listas de booleanos. Nuevamente puede parecer un tanto incoherente aunque, en este caso, su uso sí es extremadamente conveniente. Veamos por qué:

Si partimos del mismo dataframe usado en la sección anterior, podemos crear una lista de booleanos (que, por motivos puramente pedagógicos, asignamos a una variable, mask) y realizar la selección con ella entre los corchetes. Vemos a continuación que este método también selecciona filas del dataframe:

In [214]:
df = pd.DataFrame(np.arange(18).reshape([6, 3]),
                 index = ["a", "b", "c", "d", "e", "f"],
                 columns = ["A", "B", "C"])
df

A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17

In [215]:
mask = [True, False, True, False, False, True] #Muestra los indices que cumplen con True
df[mask]

A   B   C
a   0   1   2
c   6   7   8
f  15  16  17

El vector de booleanos deberá tener la misma longitud que el índice de filas (es decir, un booleano por fila) y la selección devolverá aquellas filas para las que el elemento correspondiente del vector tome el valor True.

La verdadera potencia de este estilo de selección se pone de manifiesto cuando la máscara se genera a partir de los datos del propio dataframe. Por ejemplo, si queremos seleccionar las filas para las que el valor de la columna A sea mayor que 7:

In [216]:
df[df.A > 7] #Muestra los indices que cumplen con la condición

A   B   C
d   9  10  11
e  12  13  14
f  15  16  17

Este tipo de filtrados resultan muy frecuentes en entornos de análisis, de ahí que la posibilidad de realizarlos sin necesidad de recurrir a métodos adicionales (loc, iloc o get, por ejemplo) resulte tan conveniente.

Aun así, esta técnica también es compatible con los métodos loc e iloc, con algún matiz adicional:

Con loc podemos usar directamente una expresión de comparación como la vista:

In [217]:
df.loc[df.B > 6] #Muestra los indices que cumplen con la condición

A   B   C
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17

Sin embargo, con iloc nos veremos obligados a extraer los valores del dataframe resultante de la comparación -tal y como ocurría con las series- pues, de otro modo, obtendremos un error:

In [218]:
df.iloc[(df.B > 6 ).values]

A   B   C
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17

Evitamos problemas si, tal y como sugiere pandas, utilizamos siempre el método loc.

# Selección aleatoria

Al igual que ocurre con las series, también los dataframes tienen un método que permite extraer elementos del mismo de forma aleatoria: pandas.DataFrame.sample. Este método permite especificar el número de elementos a extraer (o el porcentaje respecto del total, parámetros **n** y **frac**, respectivamente), si la extracción se realiza con reemplazo o no (parámetro **replace**), los pesos a aplicar a los elementos para realizar una extracción aleatoria ponderada (parámetro **weights**) y una semilla para el generador de números aleatorios que asegure la reproducibilidad de la extracción (parámetro **random_state**). También es posible indicar el eje a lo largo del cual se desea realizar la extracción (por defecto se extraen filas, correspondiente al eje 0).

Veamos un ejemplo. Si partimos del siguiente dataframe:

In [219]:
df = pd.DataFrame(np.arange(18).reshape([6, 3]),
                  index = ["a", "b", "c", "d", "e", "f"],
                  columns = ["A", "B", "C"])
df

A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  12  13  14
f  15  16  17

podemos extraer 3 filas de forma aleatoria, sin reemplazo (opción por defecto) y fijando como semilla del generador de números aleatorios el número 18, de la siguiente forma:

In [220]:
df.sample(2, random_state = 18)

A   B   C
f  15  16  17
e  12  13  14

Si especificamos como eje el valor 1, estaremos extrayendo columnas:

In [221]:
df.sample(2, random_state = 18, axis = 1) #Columnas

A   B
a   0   1
b   3   4
c   6   7
d   9  10
e  12  13
f  15  16

Si hacemos uso del parámetro frac, podemos especificar el porcentaje de elementos a extraer:

In [ ]:
df.sample(frac = 0.6, random_state = 18)

# El método pop

Otra forma de extraer datos es la proporcionada por el método pandas.DataFrame.pop, que extrae y elimina una columna de un dataframe:

In [ ]:
df = pd.DataFrame(np.arange(15).reshape([3, 5]), 
                  index = ["a", "b", "c"],
                  columns = ["A", "B", "C", "D", "E"])
df

In [ ]:
columna = df.pop("B") #Elimina la columna "B"
columna

In [ ]:
df

# Resumen de tipos de selección 

La cantidad de posibilidades que nos ofrece la librería pandas a la hora de extraer información de un dataframe puede resultar un tanto abrumadora. Veamos un resumen de las principales técnicas vistas y qué estamos extrayendo con cada una. En la siguiente lista e representa una etiqueta, i un índice y b un booleano:

<img src="https://i.imgur.com/B0BpStV.png">

No se incluyen en el listado anterior combinaciones de estos métodos ni los métodos para la selección usando índices y etiquetas simultáneamente.

Aunque, a primera vista, pueda parecer bastante confuso, es posible destacar algunas reglas básicas:

1. Cuando se usan los métodos loc o iloc, el primer argumento siempre hace referencia a filas y el segundo a columnas. Esto significa que si no se incluye el segundo argumento, siempre estaremos extrayendo filas.
2. El método get devuelve columnas.
3. Sin incluir los métodos loc, iloc y get, solo hay -en el listado anterior- dos formas de extraer columnas: usando como argumento una etiqueta y usando como argumento una lista de etiquetas.
4. Del punto anterior extraemos como corolario que cualquier otra notación va a devolver filas (el uso de rangos y el uso de listas de booleanos)

Y, de hecho, si de los cuatro puntos anteriores quitamos los dos primeros por obvios, nos quedan dos reglas muy simples:

1. El uso de una etiqueta o de una lista de etiquetas devuelve columnas
2. En otros casos se devuelven filas (rangos de números o de etiquetas, o listas de booleanos)

Y, al respecto del tipo de estructura devuelta (si es un escalar, una serie o un dataframe) también es posible identificar una regla: salvo el caso más obvio en el que estemos extrayendo un valor resultante de la intersección de una fila y una columna, si la nomenclatura que estamos usado permite extraer más de una fila o de una columna (aunque estemos extrayendo solo una en un momento dado) devolverá siempre dataframes y, en caso contrario, series.

Por ejemplo, el uso de rangos permite extraer más de una fila (o de una columna), de forma que su uso siempre devuelve un dataframe. Así, df["e1":"e2"] siempre devolverá un dataframe, aunque en este ejemplo estemos usando el rango para extraer una única columna.

Otro ejemplo: si estamos usando la notación df.loc[:, "e"], estamos extrayendo una columna y solo una columna, y con esta notación (dos puntos para las filas y una etiqueta para columnas) nunca podríamos extraer más de una columna, de forma que el resultado de la extracción siempre será una serie.

# Edición de series 

Podemos modificar un valor de una serie usando la notación corchetes, y haciendo referencia a índices o a etiquetas:

In [222]:
s = pd.Series([1, 2, 3, 4, 5], index = ["a", "b", "c", "d", "e"])
s

a    1
b    2
c    3
d    4
e    5
dtype: int64

In [223]:
s[0] = -1

In [224]:
s["b"] = -2

In [225]:
s

a   -1
b   -2
c    3
d    4
e    5
dtype: int64

Podemos asignar un valor a un rango, definido éste por índices o por etiquetas, asignándose dicho valor a cada uno de los elementos involucrados en el rango:

In [227]:
s[1:3] = 0
s

a   -1
b    0
c    0
d   -2
e    5
dtype: int64

In [226]:
s["b":"d"] = -2 
s

a   -1
b   -2
c   -2
d   -2
e    5
dtype: int64

Como ya hemos visto en más de una ocasión, si el rango está delimitado por números (haciendo referencia a la posición de los elementos), el último elemento del rango no se incluye. Por el contrario, si está delimitado por etiquetas, el último elemento sí se incluye.

Al rango podemos asignar también una lista de valores, aunque en este caso la lista deberá tener el mismo número de elementos que el rango en cuestión:

In [228]:
s[1:3] = [0, 1] 
s

a   -1
b    0
c    1
d   -2
e    5
dtype: int64

In [229]:
s = pd.Series([1, 2, 3, 4, 5], index = ["a", "b", "c", "d", "e"])
s
s["b":"d"] = [10, 11, 12]
s

a     1
b    10
c    11
d    12
e     5
dtype: int64

Si asignamos un valor haciendo referencia a una etiqueta que no existe en el índice, se añade dicha etiqueta al índice y se le asigna el valor:

In [230]:
s["f"] = 0 #Se añade "f"
s

a     1
b    10
c    11
d    12
e     5
f     0
dtype: int64

Esto solo funciona con etiquetas. Si utilizamos un índice y éste no existe en la serie, se devolverá un error.

Si usamos un rango con etiquetas y alguna de las etiquetas no existe, solo se asigna el valor a las existentes:

In [231]:
s["f":"h"] = 99 #Solo cambia el valor de "f", "h" no existe
s

a     1
b    10
c    11
d    12
e     5
f    99
dtype: int64

Por último, también podemos usar en la selección una lista -tanto de índices como de etiquetas-, en cuyo caso ya sabemos que estamos seleccionando los valores indicados en el orden indicado. Por ejemplo, podemos usar la lista ["c", "a"] para asignar a los elementos correspondientes los valores 1 y 2, respectivamente:

In [232]:
s[["c", "d"]] = [100, 200] #Esto es una lista, no un indice
s

a      1
b     10
c    100
d    200
e      5
f     99
dtype: int64

Si utilizamos índices, el resultado es semejante:

In [233]:
s[[1, 0]] = [2000, 3000] #Esto es una lista, no un indice
s

a    3000
b    2000
c     100
d     200
e       5
f      99
dtype: int64

# Eliminación series método pop y drop

El método pop ya ha sido presentado por eso se presentará el método pandas.Series.drop el cual devuelve una copia de la serie tras eliminar el elemento cuya etiqueta se especifica como argumento:

In [234]:
s = pd.Series([1, 2, 3, 4, 5], 
              index = ["a", "b", "c", "d", "e"])
s

a    1
b    2
c    3
d    4
e    5
dtype: int64

In [235]:
r = s.drop("b")
r

a    1
c    3
d    4
e    5
dtype: int64

En este ejemplo hemos pasado como único argumento la etiqueta del elemento a eliminar, y el método ha devuelto la serie sin dicho elemento. Si la etiqueta no se encontrase en la serie, se devolvería un error.

También podemos pasar como argumento no una etiqueta, sino una lista de etiquetas. En este caso se eliminarán todos los elementos con dichas etiquetas:

In [236]:
r = s.drop(["d", "a"])
r

b    2
c    3
e    5
dtype: int64

Obsérvese que las etiquetas no tienen que estar en orden.

El argumento inplace = True realiza la eliminación inplace (modificando directamente la serie).

Este método exige el uso de etiquetas para seleccionar los elementos a eliminar. Esto significa que si en un momento dado necesitamos eliminar uno o más elementos por su índice, deberemos convertirlos en sus correspondientes etiquetas, lo que resulta extremadamente sencillo seleccionando los elementos adecuados del index. En el siguiente ejemplo, partimos del mismo ejemplo ya visto anteriormente:

In [237]:
s = pd.Series([1, 2, 3, 4, 5], index = ["a", "b", "c", "d", "e"])
s

a    1
b    2
c    3
d    4
e    5
dtype: int64

Si quisiéramos eliminar los elementos cuyos índices son 1 y 3, bastaría recordar que el atributo index devuelve todas las etiquetas y que s.index[[1, 3]] devuelve las correspondientes a dichos índices:

In [238]:
s.index[[1, 3]]

Index(['b', 'd'], dtype='object')

Si pasamos esta expresión como argumento del método drop, obtendremos el resultado esperado:

In [239]:
s.drop(s.index[[1, 3]])

a    1
c    3
e    5
dtype: int64

# Método Where (Buscar - Buscar/Reemplazar)

El método pandas.Series.where permite filtrar los valores de una serie de forma que solo los que cumplan cierta condición se mantengan. Los valores que no la cumplan son sustituidos por un valor (NaN por defecto):

In [240]:
s = pd.Series(np.arange(0, 10))
s

0    0
1    1
2    2
3    3
4    4
5    5
6    6
7    7
8    8
9    9
dtype: int32

Supongamos ahora que queremos filtrar los valores de s que sean pares:

In [241]:
s.where(s % 2 == 0) #Where devuelve los valores NaN para quienes no cumplen con la condición

0    0.0
1    NaN
2    2.0
3    NaN
4    4.0
5    NaN
6    6.0
7    NaN
8    8.0
9    NaN
dtype: float64

Comprobamos que los valores que no cumplen la condición son sustituidos por NaN. Podemos modificar este valor de reemplazo pasando al método como segundo argumento el valor que queremos fijar:

In [242]:
s.where(s % 2 == 0, -1) #Con el segundo argumento, reemplaza os valores

0    0
1   -1
2    2
3   -1
4    4
5   -1
6    6
7   -1
8    8
9   -1
dtype: int32

# Edición de DataFrame

Hemos visto la gran variedad de formas que tenemos a nuestra disposición para seleccionar elementos o bloques de elementos de un dataframe, y cada una de estas selecciones puede ser utilizada para modificar los valores contenidos en el dataframe. Veamos algunos ejemplos:

Podemos modificar un valor concreto usando los métodos loc o iloc, en función de que queramos usar sus etiquetas o índices:

In [243]:
df = pd.DataFrame(np.arange(12).reshape([4, 3]), 
                  index = ["a", "b", "c", "d"],
                  columns = ["A", "B", "C"])
df

A   B   C
a  0   1   2
b  3   4   5
c  6   7   8
d  9  10  11

In [244]:
df.iloc[1, 2] = -1
df

A   B   C
a  0   1   2
b  3   4  -1
c  6   7   8
d  9  10  11

Podemos modificar una columna completa seleccionándola y asignándole, por ejemplo, una lista con los nuevos valores. Si partimos del mismo ejemplo que en el caso anterior...

In [245]:
df = pd.DataFrame(np.arange(12).reshape([4, 3]), 
                  index = ["a", "b", "c", "d"],
                  columns = ["A", "B", "C"])
print(df)

df["A"] = [10, 20, 30, 40]
df

   A   B   C
a  0   1   2
b  3   4   5
c  6   7   8
d  9  10  11


A   B   C
a  10   1   2
b  20   4   5
c  30   7   8
d  40  10  11

En este caso, la longitud de la lista conteniendo los valores a insertar deberá coincidir con la longitud de la columna, salvo que en lugar de una lista se esté asignando un único valor, en cuyo caso se propagará a toda la columna.

Si la selección es un bloque de datos de un tamaño arbitrario, nos encontramos en el mismo escenario: o bien insertamos datos con el mismo tamaño que la selección, o insertamos un único valor que se propagará a toda la selección. Veamos el primer caso:

In [246]:
df = pd.DataFrame(np.arange(12).reshape([4, 3]), 
                  index = ["a", "b", "c", "d"],
                  columns = ["A", "B", "C"])
print(df)

df.loc["b":"c", "A":"B"] = [[-1, -2], [-3, -4]] # [F, C]
df

   A   B   C
a  0   1   2
b  3   4   5
c  6   7   8
d  9  10  11


A   B   C
a  0   1   2
b -1  -2   5
c -3  -4   8
d  9  10  11

En este ejemplo hemos seleccionado un bloque de 2x2, y hemos insertado datos con una estructura de las mismas dimensiones.

In [247]:
df = pd.DataFrame(np.arange(12).reshape([4, 3]), 
                  index = ["a", "b", "c", "d"],
                  columns = ["A", "B", "C"])
print(df)

df.loc["b":"c", "A":"B"] = -1 
df

   A   B   C
a  0   1   2
b  3   4   5
c  6   7   8
d  9  10  11


A   B   C
a  0   1   2
b -1  -1   5
c -1  -1   8
d  9  10  11

y en este segundo caso hemos asignado un único valor a la misma selección.

Nos hemos encontrado también con el caso de insertar datos en una columna o fila inexistente, en cuyo caso se crea y se le asignan los valores en cuestión. En el primer caso (de columna inexistente):

In [248]:
df = pd.DataFrame(np.arange(12).reshape([4, 3]), 
                  index = ["a", "b", "c", "d"],
                  columns = ["A", "B", "C"])
print(df)

df["D"] = [10, 20, 30, 40] 
df

   A   B   C
a  0   1   2
b  3   4   5
c  6   7   8
d  9  10  11


A   B   C   D
a  0   1   2  10
b  3   4   5  20
c  6   7   8  30
d  9  10  11  40

In [ ]:
Y en el segundo (de fila inexistente):

In [249]:
df = pd.DataFrame(np.arange(12).reshape([4, 3]), 
                  index = ["a", "b", "c", "d"],
                  columns = ["A", "B", "C"])
print(df)

df.loc["e"] = [10, 20, 30] 
df

   A   B   C
a  0   1   2
b  3   4   5
c  6   7   8
d  9  10  11


A   B   C
a   0   1   2
b   3   4   5
c   6   7   8
d   9  10  11
e  10  20  30

# El método Where

De forma semejante a las series, el método de los dataframes where filtra los valores contenidos en el dataframe de forma que solo los que cumplan cierta condición se mantengan. El resto de valores son sustituidos por un valor que, por defecto, es NaN.

Por ejemplo, partimos del siguiente dataframe:

In [250]:
df = pd.DataFrame(np.arange(12).reshape(-1, 3), columns=["A", "B", "C"]) #-1 es igual a 0
df

A   B   C
0  0   1   2
1  3   4   5
2  6   7   8
3  9  10  11

Si ahora queremos filtrar los valores múltiplos de 2, por ejemplo, podemos hacerlo de la siguiente forma:

In [251]:
df.where(df % 2 == 0)

A     B    C
0  0.0   NaN  2.0
1  NaN   4.0  NaN
2  6.0   NaN  8.0
3  NaN  10.0  NaN

Todos aquellos valores que no son múltiplo de 2 son sustituidos por NaN. Si, por ejemplo, quisiéramos cambiar de signo a los valores que no cumplen la condición impuesta, lo haríamos así:

In [252]:
df = pd.DataFrame(np.arange(12).reshape(-1, 3), columns=["A", "B", "C"]) #-1 es igual a 0
print(df)

df.where(df % 2 == 0, -df) #Cambiamos el signo desde la matriz "df"

   A   B   C
0  0   1   2
1  3   4   5
2  6   7   8
3  9  10  11


A   B   C
0  0  -1   2
1 -3   4  -5
2  6  -7   8
3 -9  10 -11

# Eliminación de elementos 

El método pandas.DataFrame.drop elimina las filas o columnas indicadas y devuelve el resultado, permitiéndose diferentes criterios para especificarlas.

El primer criterio consiste en indicar la lista de etiquetas a eliminar y el eje al que pertenecen. Partamos del siguiente dataframe:

In [253]:
df = pd.DataFrame(np.arange(16).reshape([4, 4]), 
                  index = ["a", "b", "c", "d"],
                  columns = ["A", "B", "C", "D"])
df

A   B   C   D
a   0   1   2   3
b   4   5   6   7
c   8   9  10  11
d  12  13  14  15

Podemos eliminar, por ejemplo, las filas cuyas etiquetas son "a" y "c" con el siguiente código:

In [254]:
df.drop(["a", "c"],axis = 0)  #axis 0 Filas, axis 1 columnas

A   B   C   D
b   4   5   6   7
d  12  13  14  15

Obsérvese que lo que se muestra es el resultado de eliminar las filas indicadas del dataframe. Éste no se modifica salvo que utilicemos el argumento inplace = True.

Como el eje por defecto es el 0, la instrucción anterior es equivalente a:

In [255]:
df.drop(["a", "c"]) #axis 0 Filas, axis 1 columnas

A   B   C   D
b   4   5   6   7
d  12  13  14  15

Para eliminar columnas, habría que indicar el eje correspondiente:

In [256]:
df.drop(["B", "D"], axis = 1) #axis 0 Filas, axis 1 columnas

A   C
a   0   2
b   4   6
c   8  10
d  12  14

Otra alternativa para especificar si estamos eliminando filas o columnas es utilizar directamente los parámetros index y columns. Así, otra forma de eliminar las filas "a" y "c" sería la siguiente:

In [257]:
df.drop(index = ["a", "c"])

A   B   C   D
b   4   5   6   7
d  12  13  14  15

-el resultado es el mismo que antes, lógicamente-. Y para eliminar las columnas "B" y "D":

In [258]:
df.drop(columns = ["B", "D"])

A   C
a   0   2
b   4   6
c   8  10
d  12  14

# Unión de series y dataframes

Frecuentemente nos encontramos con que los datos a analizar están repartidos entre dos o más bloques de datos, lo que nos obliga a unirlos, bien concatenándolos, o bien realizando un "join" entre las estructuras (uniones del mismo tipo que las realizadas en bases de datos). Revisemos las funciones asociadas.

# Unión de series

Comencemos revisando las opciones disponibles para las series pandas

# La función concat

Un caso con el que nos encontramos con relativa frecuencia es aquel en el que queremos unir una serie a otra. Por ejemplo:

In [259]:
s = pd.Series([1, 2, 3, 4, 5,], index = ["a", "b", "c", "d", "e"])
r = pd.Series([10, 11, 12], index = ["f", "g", "h"])

Si deseamos unir r y s en una nueva serie, podemos usar la función pandas.concat. Esta función permite especificar el eje a lo largo del cual unir los diferentes objetos (pueden ser series o dataframes). Por defecto, la concatenación se realiza a lo largo del eje 0:

In [260]:
t = pd.concat([s, r])
print(type(t))
t

<class 'pandas.core.series.Series'>


a     1
b     2
c     3
d     4
e     5
f    10
g    11
h    12
dtype: int64

Podemos ver en el ejemplo anterior que el resultado es una serie pandas.

Si especificamos como eje de concatenación el eje 1, pandas alineará los valores con idénticas etiquetas. En el siguiente ejemplo, las series a y b tienen algunas etiquetas comunes (y otras no). El resultado incluye todas las etiquetas asignando el valor NaN ("Not a Number") a aquellos valores desconocidos:

In [261]:
a = pd.Series([1, 2, 3, 4, 5,], index = ["a", "b", "c", "d", "e"])
b = pd.Series([10, 11, 12], index = ["a", "b", "f"])
pd.concat([a, b], axis = 1, sort = True)

0     1
a  1.0  10.0
b  2.0  11.0
c  3.0   NaN
d  4.0   NaN
e  5.0   NaN
f  NaN  12.0

(se ha utilizado el argumento sort = True para ocultar cierto aviso al respecto de un cambio en la funcionalidad de esta función en versiones futuras de la librería pandas)

Como puede observarse, el resultado es un dataframe:

In [262]:
type(pd.concat([a, b], axis = 1, sort = True))

pandas.core.frame.DataFrame

Por otro lado, ya sabemos que las etiquetas del índice no tienen por qué ser diferentes, de forma que si estuviésemos concatenando series con etiquetas comunes en sus índices, el resultado sería equivalente a los vistos hasta ahora:

In [263]:
s = pd.Series([1, 2, 3, 4], index = ["a", "b", "c", "d"])
r = pd.Series([10, 11, 12], index = ["a", "c", "e"])
pd.concat([s, r])

a     1
b     2
c     3
d     4
a    10
c    11
e    12
dtype: int64

En este ejemplo hemos concatenado dos series que tienen dos etiquetas comunes ("a" y "c"), y vemos que las dos apariciones de cada una de ellas se incluyen en el resultado de la concatenación.

# El método append

Otra alternativa es usar el método pandas.Series.append, versión simplificada de la función concat ya vista que devuelve la unión de la serie sobre la que se aplica con otra (u otras) series, pero solo a lo largo del eje 0. Veámoslo en funcionamiento:

In [264]:
#Version concat simplificada
a = pd.Series([1, 2, 3, 4, 5], index = ["a", "b", "c", "d", "e"])
b = pd.Series([10, 11, 12], index = ["f", "g", "h"])

In [265]:
c = a.append(b)
c

a     1
b     2
c     3
d     4
e     5
f    10
g    11
h    12
dtype: int64

Si el argumento ignore_index toma el valor True, se ignoran las etiquetas de las series:

In [266]:
c = a.append(b, ignore_index = True) #Solo valores
c

0     1
1     2
2     3
3     4
4     5
5    10
6    11
7    12
dtype: int64

# Concatenación y unión de dataframes

Ésta es otra de las áreas en las que la variedad de opciones puede resultar confusa. A modo de resumen, digamos que pandas ofrece dos principales funciones con este objetivo: pandas.concat y pandas.merge.

- La función concat permite concatenar dataframes a lo largo de un determinado eje
- La función merge permite realizar uniones (joins) entre dataframes tal y como se realizan en bases de datos. Esta función también está disponible como método: pandas.DataFrame.merge

Hay una tercera función que está disponible solo como método: pandas.DataFrame.append. El método append ofrece una funcionalidad semejante a la de la función concat pero reducida. Así, por ejemplo, solo permite realizar concatenaciones a lo largo del eje 0 (es decir, verticalmente).

Veamos en las siguientes páginas algunos ejemplos de estas funciones.

# La función concat

La función pandas.concat es la responsable de concatenar dos o más dataframes (y de todas las estructuras proveídas por pandas) a lo largo de un eje, con soporte a lógica de conjuntos a la hora de gestionar etiquetas en ejes no coincidentes. Veamos un primer caso, el más sencillo posible, para el que partimos de los siguientes dos dataframes:

In [ ]:
import numpy as np

df1 = pd.DataFrame(np.arange(9).reshape([3, 3]),
                  index = ["a", "b", "d"],
                  columns = ["A", "B", "C"]) 
df1

In [ ]:
df2 = pd.DataFrame(np.arange(12).reshape([4, 3]),
                  index = ["a", "b", "c", "e"],
                  columns = ["B", "C", "D"]) 
df2

Nota: tal y como ha ocurrido en secciones anteriores, en algunos de los ejemplos que se muestran a continuación se utiliza el argumento sort = False para evitar que se muestre un aviso al respecto de cierto cambio de funcionalidad que se producirá en futuras versiones de esta librería.

Si pasamos a la función concat ambos dataframes como primer argumento (en forma de lista), obtenemos el siguiente resultado:

In [ ]:
pd.concat([df1, df2], sort = False) 

Vemos cómo, por defecto, la concatenación se ha realizado a lo largo del eje 0 (eje vertical), uniendo los índices de fila de ambos dataframes, y alineando las columnas por su etiqueta. Los valores para los que no hay datos se han rellenado con NaN (opción correspondiente al argumento por defecto join: "outer").

Si especificamos que la concatenación se realice a lo largo del eje 1 (eje horizontal), el resultado es el siguiente:

In [ ]:
pd.concat([df1, df2], axis = 1, sort = False) # axis 0 fila, axis 1 columnas, por tanto, se agregan nuevas columnas

De modo semejante al primer ejemplo, se han introducido NaN's allí donde no había datos, y se han alineado las filas por su etiqueta.

Estos dos ejemplos vistos son tipo "Outer" (opción por defecto), considerando todas las etiquetas de los dos dataframes aun cuando no sean comunes a ambos. Pero si especificamos el argumento join = "Inner", los resultados pasan a considerar solo las etiquetas comunes. Así, para el primer ejemplo visto tenemos:

In [ ]:
pd.concat([df1, df2], join = "inner") #Solo muestra etiquetas comunes, por defecto outer.

Incluyendo solo las columnas B y C comunes a ambos dataframes. Y para el segundo ejemplo tenemos:

In [ ]:
pd.concat([df1, df2], axis = 1, join = "inner") #Solo muestra etiquetas comunes, por defecto outer.

...incluyendo solo las filas a y b comunes a ambos dataframes.

El parámetro ignore_index controla el índice a asignar al eje a lo largo del cuál se realiza la concatenación. Si este parámetro toma el valor False (por defecto), el eje de concatenación mantiene las etiquetas de los dataframes originales. Si toma el valor True, se ignoran dichas etiquetas y el resultado de la concatenación recibe un nuevo índice automático numérico. Por ejemplo, si añadimos al siguiente ejemplo el argumento ignore_index=True, obtenemos el siguiente resultado:

In [ ]:
pd.concat([df1, df2], axis = 1, join = "inner", ignore_index = True)

# El método append

Como se ha comentado, el método pandas.DataFrame.append es un atajo de la función concat que ofrece funcionalidad semejante pero limitada: no permite especificar el eje de concatenación (siempre es el eje 0) ni el tipo de "join" (siempre es tipo "Outer").

Si seguimos con los mismos dataframes de la sección anterior:

In [ ]:
df1 = pd.DataFrame(np.arange(9).reshape([3, 3]),
                  index = ["a", "b", "d"],
                  columns = ["A", "B", "C"]) 
df1

In [ ]:
df2 = pd.DataFrame(np.arange(12).reshape([4, 3]),
                  index = ["a", "b", "c", "e"],
                  columns = ["B", "C", "D"]) 
df2

...podemos ver cuál es el resultado de aplicar este método a df1:

In [ ]:
df1.append(df2, sort = False)

Al igual que ocurría con la función concat, el parámetro ignore_index nos permite controlar las etiquetas que recibe el índice del resultado: las de los dataframes originales (con ignore_index = False, opción por defecto), o uno nuevo automático (con ignore_index = True).